In [1]:
import pandas as pd
from sklearn.metrics import confusion_matrix
import numpy as np
import jsonlines

In [2]:
LABELS = ['B-MUS', 'B-PER', 'I-MUS', 'I-PER', 'O']    

original format:
```
{"id": 50, "sent_items": [["A", "O"], ["large", "O"], ["ensemble", "O"], ["of", "O"], ["trombonists", "O"], ["would", "O"], ["gather", "O"], ["to", "O"], ["play", "O"], ["music", "O"], ["written", "O"], ["for", "O"], ["multiple", "O"], ["trombones", "O"], ["or", "O"], ["transcribed", "O"], ["from", "O"], ["other", "O"], ["sources", "O"], [",", "O"], ["such", "O"], ["as", "O"], ["the", "O"], ["chorales", "O"], ["of", "O"], ["Johann", "B-MUS"], ["Sebastian", "I-MUS"], ["Bach", "I-MUS"], [".", "O"]]}
```

In [13]:
def convert_dev_from_json_to_conll(dev_path, conll_path):
    with jsonlines.open(dev_path, "r") as fin:
        with open(conll_path, "w") as fout:
            for line in fin:
                tokens = line["sent_items"]
                print(tokens)
                print([f"{t[0]} {t[1]}\n" for t in tokens])
                for t in tokens:
                    fout.write(f"{t[0]} {t[1]}\n")
                    
                    
def get_set_labels(set_path):
    set_labels = []
    with open(set_path, "r") as f:
        for t in f.readlines():
            if t.strip():
                label = t.split()[-1]
                set_labels.append(label)
            else:
                print(t)
    return set_labels
                    


In [4]:
dev_path = '../data/musicians_dataset/dev_converted.jsonl'
conll_path = '../experiments/gold_devs/dev_converted.conll'


convert_dev_from_json_to_conll(dev_path, conll_path)



In [8]:
gold_path = '../experiments/gold_devs/dev_converted.conll'

pred_path = '../experiments/predictions/sents_with_no_ents_bigger-manual.conll'
gold_labels = get_set_labels(gold_path)
pred_labels = get_set_labels(pred_path)
print(set(gold_labels))
print(set(pred_labels))


confusion_matrix(gold_labels, pred_labels, labels=['B-MUS', 'B-PER', 'I-MUS', 'I-PER', 'O'])

{'B-PER', 'I-MUS', 'O', 'I-PER', 'B-MUS'}
{'B-PER', 'I-MUS', 'O', 'I-PER', 'B-MUS'}


array([[  26,   62,    0,    1,    9],
       [   0,  162,    0,    6,   11],
       [   1,    0,   28,   57,    2],
       [   0,    0,    0,  106,    8],
       [   0,   26,    4,    9, 4028]])

In [14]:
dev_path = '../data/musicians_dataset/split_dev_20211118.jsonl'
conll_path = '../experiments/gold_devs/split_dev_20211118.conll'


convert_dev_from_json_to_conll(dev_path, conll_path)

[['In', 'O'], ['the', 'O'], ['pilot', 'O'], ['episode', 'O'], ['of', 'O'], ['the', 'O'], ['sequel', 'O'], ['series', 'O'], ['"', 'O'], ['Joey', 'O'], ['"', 'O'], [',', 'O'], ['Joey', 'B-PER'], ["'s", 'O'], ['sister', 'O'], ['Gina', 'B-PER'], ['briefly', 'O'], ['expresses', 'O'], ['her', 'O'], ['belief', 'O'], ['that', 'O'], ['Joey', 'B-PER'], ['and', 'O'], ['Chandler', 'B-PER'], ['were', 'O'], ['a', 'O'], ['gay', 'O'], ['couple', 'O'], ['.', 'O']]
['In O\n', 'the O\n', 'pilot O\n', 'episode O\n', 'of O\n', 'the O\n', 'sequel O\n', 'series O\n', '" O\n', 'Joey O\n', '" O\n', ', O\n', 'Joey B-PER\n', "'s O\n", 'sister O\n', 'Gina B-PER\n', 'briefly O\n', 'expresses O\n', 'her O\n', 'belief O\n', 'that O\n', 'Joey B-PER\n', 'and O\n', 'Chandler B-PER\n', 'were O\n', 'a O\n', 'gay O\n', 'couple O\n', '. O\n']
[['It', 'O'], ['was', 'O'], ['subsequently', 'O'], ['covered', 'O'], ['by', 'O'], ['many', 'O'], ['artists', 'O'], [',', 'O'], ['most', 'O'], ['famously', 'O'], ['by', 'O'], ['Quincy'

In [16]:
len(['The O\n', 'Polish O\n', 'NREAP O\n', 'plan O\n', 'is O\n', 'targeting O\n', '6,700 O\n', 'MW O\n', 'of O\n', 'wind O\n', 'power O\n', 'by O\n', '2020 O\n', 'whilst O\n', 'EWEA O\n', "'s O\n", '2009 O\n', 'forecast O\n', 'suggests O\n', 'a O\n', 'higher O\n', 'wind O\n', 'capacity O\n', 'of O\n', '10–12 O\n', '\xa0 O\n', 'GW O\n', 'is O\n', 'possible O\n', '. O\n'])

30

In [10]:
gold_path = '../experiments/gold_devs/split_dev_20211118.conll'

pred_path = '../experiments/predictions/sents_with_no_ents_bigger-split.conll'
gold_labels = get_set_labels(gold_path)
pred_labels = get_set_labels(pred_path)
print(set(gold_labels) - set(pred_labels))
print()


confusion_matrix(gold_labels, pred_labels, labels=LABELS)

set()



ValueError: Found input variables with inconsistent numbers of samples: [7677, 7662]

In [20]:
import re

def untag_dev_sentences(devpath):
    clean_sentences = []
    tokens = []
    with open(devpath, "r") as fin:
        for line in fin.readlines():
            sentence = re.sub('\-\[[A-Z]*\]',  '', line)
            clean_sentences.append(sentence)
            for token in sentence.split():
                if token.strip():
                    tokens.append(token)
    return clean_sentences, tokens

In [21]:
clean_sentences, tokens = untag_dev_sentences("../data/musicians_dataset/split_dev_2.jsonl")

In [23]:
tokens
# len(gold_labels)

['{"id":',
 '72163589,',
 '"sent_items":',
 '[["In",',
 '"O"],',
 '["the",',
 '"O"],',
 '["three",',
 '"O"],',
 '["following",',
 '"O"],',
 '["years",',
 '"O"],',
 '["she",',
 '"O"],',
 '["participated",',
 '"O"],',
 '["again",',
 '"O"],',
 '["in",',
 '"O"],',
 '["the",',
 '"O"],',
 '["Festival",',
 '"O"],',
 '["of",',
 '"O"],',
 '["Sanremo",',
 '"O"],',
 '[",",',
 '"O"],',
 '["winning",',
 '"O"],',
 '["the",',
 '"O"],',
 '["critic",',
 '"O"],',
 '["\'s",',
 '"O"],',
 '["award",',
 '"O"],',
 '["in",',
 '"O"],',
 '["1987",',
 '"O"],',
 '["with",',
 '"O"],',
 '["the",',
 '"O"],',
 '["song",',
 '"O"],',
 '["\\"",',
 '"O"],',
 '["Primo",',
 '"B-MUS"],',
 '["tango",',
 '"O"],',
 '["\\"",',
 '"O"],',
 '["(",',
 '"O"],',
 '["First",',
 '"O"],',
 '["Tango",',
 '"O"],',
 '[")",',
 '"O"],',
 '[".",',
 '"O"]]}',
 '{"id":',
 '3119,',
 '"sent_items":',
 '[["The",',
 '"O"],',
 '["watershed",',
 '"O"],',
 '["is",',
 '"O"],',
 '["bordered",',
 '"O"],',
 '["on",',
 '"O"],',
 '["the",',
 '"O"],',
 '["so

### manual

         ['B-MUS', 'B-PER', 'I-MUS', 'I-PER', 'O'] 
['B-MUS', [  29,   61,    0,    1,    7],
 'B-PER',[   1,   158,    0,    6,   14],
 'I-MUS'[    1,    0,    32,   51,    4],
 'I-PER',[   0,    0,    0,   106,    8],
 'O']    [   1,   29,    1,   15,   4021]
 

### Slit
 
 

In [19]:
from collections import Counter

In [43]:
dev_musicians = []
with jsonlines.open('../data/musicians_dataset/split_dev_only_hearst.jsonl') as f:
    for sentence in f:
        musician = ""
        for item in sentence["sent_items"]:
            if "-MUS" in item[1]:
                musician += item[0] + " "
            elif item[1] == "O" and musician != "":
                dev_musicians.append(musician[:-1])
                musician = ""

In [45]:
train_musicians = []
with jsonlines.open('../data/musicians_dataset/split_train_only_hearst_uniques.jsonl') as f:
    for sentence in f:
        musician = ""
        for item in sentence["sent_items"]:
            if "-MUS" in item[1]:
                musician += item[0] + " "
            elif item[1] == "O" and musician != "":
                train_musicians.append(musician[:-1])
                musician = ""

In [27]:
mandev_musicians = []
with jsonlines.open('../data/musicians_dataset/dev_converted.jsonl') as f:
    for sentence in f:
        musician = ""
        for item in sentence["sent_items"]:
            if "-MUS" in item[1]:
                musician += item[0] + " "
            elif item[1] == "O" and musician != "":
                mandev_musicians.append(musician[:-1])
                musician = ""

In [30]:
print(len(mandev_musicians), len(dev_musicians))

97 148


In [33]:
dev_musicians

['Don Fleming',
 'Airto Moreira',
 'Joe Farrell',
 'Duke Ellington',
 'Elliott Sharp',
 'Martin Luther',
 'Joseph Haydn',
 'Alfred Cellier',
 'Hélio Delmiro',
 'Maeve Mackinnon',
 'Miles Davis',
 'Franciscus Henricus Antheunis',
 'Elvis Costello',
 'Elvis Costello',
 'Gary Brooker',
 'Jerry Garcia',
 'Nicky Hopkins',
 'Steve Kimock',
 'Bob Weir',
 'Mickey Hart',
 'Richard Wagner',
 'Luciano Pavarotti',
 'Bob Dylan',
 'Dean Martin',
 'Frank Sinatra',
 'Tom Jones',
 'Michael Devlin',
 'Chick Corea',
 'Michael Jackson',
 'Thomas Martin Dumont',
 'Elton John',
 'Peter Gabriel',
 'Igor Stravinsky',
 'Bonnie Tyler',
 'Elvis Presley',
 'Pharrell Williams',
 'Rashied Ali',
 'Marilyn Crispell',
 'Matthijs Vermeulen',
 'Martin Luther',
 'Pierre Boulez',
 'Billy Joel',
 'Johanna Nurmimaa',
 'Warren Ellis',
 'Martin Luther',
 'Wynton Marsalis',
 'Luke Brandon',
 'Charles Aznavour',
 'Frank Sinatra',
 'Amélie Veille',
 'Jeff Anderson',
 'Martha Wash',
 'B.B. King',
 'Bernie Tormé',
 'Johann Sebasti

In [46]:
Counter(train_musicians).most_common(100)

[('Barbara Casini', 1),
 ('Marcus Byrne', 1),
 ('Landon Ronald', 1),
 ('Ramsey Lewis', 1),
 ('Jack Wilson', 1),
 ('Megan Washington', 1),
 ('Criss Oliva', 1),
 ('John Fayette Frame', 1),
 ('Bill Janovitz', 1),
 ('Maria Taylor', 1),
 ('Omar Jeffery Pineiro', 1),
 ('Joey Ramone', 1),
 ('Richard Bennett', 1),
 ('Wynton Marsalis', 1),
 ('Gus Cannon', 1),
 ('Paul Della Pelle', 1),
 ('Pepe Aguilar', 1),
 ('The Nicholas brothers', 1),
 ('Donald Nally', 1),
 ('Andreas Skår Winther', 1),
 ('Antonique Smith', 1),
 ('David Lowe', 1),
 ('Robert Kyagulanyi Ssentamu', 1),
 ('Fateh Ali Khan', 1),
 ('Charles Segal', 1),
 ('Al Foster', 1),
 ('Willem Mengelberg', 1),
 ('Alexander von Zemlinsky', 1),
 ('Oliver Korte', 1),
 ('Al Kooper', 1),
 ('Sigur Rós', 1),
 ("Colby O'Donis Colón", 1),
 ('Anne - Marie David', 1),
 ('Gennady Provatorov', 1),
 ('Guillermo Cides', 1),
 ('Jarvis Cocker of Pulp', 1),
 ('Issei Noro', 1),
 ('Albert Dadon AM', 1),
 ('Geddy Lee', 1),
 ('Rob Brown', 1),
 ('Joseph Hoffman', 1),
 

In [37]:
repetitions_in_dev = 0
repetitions_in_train = 0

mandev_musicians_set = set(mandev_musicians)
for x in Counter(train_musicians).most_common(100):
    if x[0] in mandev_musicians_set:
        print(x)
        repetitions_in_train += x[1]
        repetitions_in_dev += mandev_musicians.count(x[0])

print(repetitions_in_dev, repetitions_in_train)

('Paul McCartney', 63)
('Duke Ellington', 47)
('Johann Sebastian Bach', 40)
('Jools Holland', 22)
('Ella Fitzgerald', 22)
('Dave Matthews', 17)
6 211


In [39]:
print(len(dev_musicians), len(mandev_musicians))

148 97


In [44]:
repetitions_in_dev = 0
repetitions_in_train = 0

dev_musicians_set = set(dev_musicians)
for x in Counter(train_musicians).most_common(100):
    if x[0] in dev_musicians_set:
        print(x)
        repetitions_in_train += x[1]
        repetitions_in_dev += dev_musicians.count(x[0])

print(repetitions_in_dev, repetitions_in_train)

('Martin Luther', 143)
('Michael Jackson', 101)
('Elvis Presley', 92)
('Bob Dylan', 88)
('Paul McCartney', 63)
('Frank Sinatra', 58)
('Elton John', 49)
('Jennifer Lopez', 48)
('Duke Ellington', 47)
('Richard Wagner', 42)
('Johann Sebastian Bach', 40)
('Eric Clapton', 39)
('Miles Davis', 38)
('Britney Spears', 37)
('Stevie Wonder', 34)
('Celine Dion', 33)
('Dean Martin', 31)
('James Arthur', 30)
('John Coltrane', 27)
('Billy Joel', 24)
('Whitney Houston', 24)
('Elvis Costello', 23)
('Ray Charles', 22)
('John Cage', 22)
('Robert Schumann', 22)
('Buddy Holly', 21)
('Joseph Haydn', 21)
('Kelly Clarkson', 20)
('Henry Santos', 19)
('Chick Corea', 19)
('Tom Jones', 19)
('Muddy Waters', 19)
('James Taylor', 18)
('B.B. King', 18)
('Leonard Cohen', 17)
('Peter Gabriel', 16)
50 1384


In [42]:
len(dev_musicians)

79

In [26]:
dev_musicians_set = set(dev_musicians)
for x in Counter(train_musicians).most_common(100):
    if x[0] in dev_musicians_set:
        print(x)

('Martin Luther', 143)
('Michael Jackson', 101)
('Elvis Presley', 92)
('Bob Dylan', 88)
('Paul McCartney', 63)
('Frank Sinatra', 58)
('Elton John', 49)
('Jennifer Lopez', 48)
('Duke Ellington', 47)
('Richard Wagner', 42)
('Johann Sebastian Bach', 40)
('Eric Clapton', 39)
('Miles Davis', 38)
('Britney Spears', 37)
('Stevie Wonder', 34)
('Celine Dion', 33)
('Dean Martin', 31)
('James Arthur', 30)
('John Coltrane', 27)
('Billy Joel', 24)
('Whitney Houston', 24)
('Elvis Costello', 23)
('Ray Charles', 22)
('John Cage', 22)
('Robert Schumann', 22)
('Buddy Holly', 21)
('Joseph Haydn', 21)
('Kelly Clarkson', 20)
('Henry Santos', 19)
('Chick Corea', 19)
('Tom Jones', 19)
('Muddy Waters', 19)
('James Taylor', 18)
('B.B. King', 18)
('Leonard Cohen', 17)
('Peter Gabriel', 16)


In [32]:
len([('Martin Luther', 143),
('Michael Jackson', 101),
('Elvis Presley', 92),
('Bob Dylan', 88),
('Paul McCartney', 63),
('Frank Sinatra', 58),
('Elton John', 49),
('Jennifer Lopez', 48),
('Duke Ellington', 47),
('Richard Wagner', 42),
('Johann Sebastian Bach', 40),
('Eric Clapton', 39),
('Miles Davis', 38),
('Britney Spears', 37),
('Stevie Wonder', 34),
('Celine Dion', 33),
('Dean Martin', 31),
('James Arthur', 30),
('John Coltrane', 27),
('Billy Joel', 24),
('Whitney Houston', 24),
('Elvis Costello', 23),
('Ray Charles', 22),
('John Cage', 22),
('Robert Schumann', 22),
('Buddy Holly', 21),
('Joseph Haydn', 21),
('Kelly Clarkson', 20),
('Henry Santos', 19),
('Chick Corea', 19),
('Tom Jones', 19),
('Muddy Waters', 19),
('James Taylor', 18),
('B.B. King', 18),
('Leonard Cohen', 17),
('Peter Gabriel', 16)])

36